### PR TITLE
Keypair auth support

### PR DESCRIPTION
**Pull request by SimbaGithub**
Simba ticket 00389231
Add support for key pair authentication for PHP. Since PHP is using libsnowflakeclient the feature needs to be added here.

### CODE REVIEW

File: cpp/lib/Authenticator.cpp [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+    {
+      return 0;
+    }
+  }
+
+  SF_STATUS STDCALL auth_authenticate(SF_CONNECT * conn)
+  {
+    if (!conn || !conn->auth_object)
+    {
+      return SF_STATUS_SUCCESS;
```
Comment (sfc-gh-hchaturvedi) [inline cpp/lib/Authenticator.cpp:106]: Should we be returning success in this case?

File: cpp/lib/Authenticator.cpp [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+  {
+    switch(type) {
+      case AUTH_JWT:
+        return new AuthenticatorJWT(conn);
+      default:
+        return nullptr;
+    }
+  }
+
+  void IAuthenticator::renewDataMap(cJSON *dataMap)
```
Comment (sfc-gh-hchaturvedi) [inline cpp/lib/Authenticator.cpp:204]: @SimbaGithub can you please remind me under what scenario we'd need this?

Comment (sfc-gh-hchaturvedi) [reply cpp/lib/Authenticator.cpp:204]: Nvm .. I remember this is for renewing Token correct?

File: lib/http_perform.c [hunk 3]
```diff
@@ -331,6 +352,15 @@ sf_bool STDCALL http_perform(CURL *curl,
 
         log_trace("Running curl call");
         res = curl_easy_perform(curl);
+
+        // forcely trigger renew timeout on the first attempt
+        if ((renew_injection) && (renew_timeout > 0) &&
+            elapsed_time && (*elapsed_time <= 0))
+        {
+            my_sleep_ms(renew_timeout * 1000);
+            res = CURLE_OPERATION_TIMEOUTED;
```
Comment (sfc-gh-hchaturvedi) [inline lib/http_perform.c:361]: nit: *TIMEDOUT

**Review by sfc-gh-hchaturvedi — APPROVED** [review]
LGTM

File: cpp/lib/Authenticator.cpp (hunk 2 above)
Comment (SimbaGithub) [reply cpp/lib/Authenticator.cpp:204]: Yes, it's for renewing jwt token.

File: cpp/lib/Authenticator.cpp (hunk 1 above)
Comment (SimbaGithub) [reply cpp/lib/Authenticator.cpp:106]: I don't think !conn would happen but !conn->auth_object is a normal case that extra authenticator is not needed, like user/password or oauth.

File: lib/http_perform.c (hunk 3 above)
Comment (SimbaGithub) [reply lib/http_perform.c:361]: Will fix. Turns out both of them are defined in curl.h.

**Pull request merged**